chain/xrp/tx/tx.go

```diff
@@ -0,0 +1,228 @@
+)
+
+type TransactionType string
+
+type XRPTransaction struct {
+	Account            xc.Address       `json:"Account"`
+	Amount             AmountBlockchain `json:"Amount"`
+	Destination        xc.Address       `json:"Destination"`
+	Fee                string           `json:"Fee"`
+	Flags              int64            `json:"Flags,omitempty"`
+	LastLedgerSequence int64            `json:"LastLedgerSequence"`
+	Sequence           int64            `json:"Sequence"`
+	SigningPubKey      string           `json:"SigningPubKey"`
+	TransactionType    TransactionType  `json:"TransactionType"`
+	TxnSignature       string           `json:"TxnSignature"`
+}
+
+type AmountBlockchain struct {
+	StringValue string  `json:"-"`
+	AmountValue *Amount `json:"-"`
+	IsString    bool    `json:"-"`
+}
+
+type Amount struct {
+	Currency string `json:"currency"`
+	Issuer   string `json:"issuer"`
+	Value    string `json:"value"`
+}
+
+// Tx for Template
+type Tx struct {
+	XRPTx                *XRPTransaction
+	SignPubKey           []byte
+	EncodeForSigning     []byte
+	EncodeXRPTx          *string
+	TransactionSignature []xc.TxSignature
+}
+
+var _ xc.Tx = &Tx{}
+
+// Hash returns the tx hash or id
+func (tx Tx) Hash() xc.TxHash {
+	serializedTxInput, err := tx.Serialize()
+	if err != nil {
+		return xc.TxHash("")
+	}
+
+	encodeTxWithPrefix := TRANSACTION_HASH_PREFIX + string(serializedTxInput)
+
+	decodedBytes, err := hex.DecodeString(encodeTxWithPrefix)
+	if err != nil {
+		return xc.TxHash("")
+	}
+
+	hash := sha512.Sum512(decodedBytes)
+	firstHalf := hash[:32]
+	hashHex := hex.EncodeToString(firstHalf)
+
+	return xc.TxHash(hashHex)
+}
+
+// Sighashes returns the tx payload to sign, aka sighash
+func (tx Tx) Sighashes() ([]xc.TxDataToSign, error) {
+	if tx.XRPTx == nil {
+		return nil, errors.New("missing XRP transaction")
+	}
+
+	if tx.EncodeForSigning == nil {
+		return nil, errors.New("missing serialised XRP transaction")
+	}
+
+	// For k256 signing, XRP uses sha512[:32]
+	// https://github.com/XRPLF/xrpl-py/blob/17aad31f77452d30917b9e4544c9c87c274c0e3d/xrpl/core/keypairs/secp256k1.py#L95
+	digestSha512 := sha512.Sum512(tx.EncodeForSigning)
+	firstHalf := digestSha512[:32]
+
+	return []xc.TxDataToSign{firstHalf[:]}, nil
+
+}
+
+// AddSignatures adds a signature to Tx
+func (tx *Tx) AddSignatures(signatures ...xc.TxSignature) error {
+	if tx.TransactionSignature != nil {
+		return errors.New("transaction already signed")
+	}
+
+	for _, rsvBytes := range signatures {
+		r, s, err := DecodeEcdsaSignature(rsvBytes)
+		if err != nil {
+			return err
+		}
+
+		signature := ecdsa.NewSignature(&r, &s)
+		signatureBytes := signature.Serialize()
+		signatureHex := hex.EncodeToString(signatureBytes)
+		tx.XRPTx.TxnSignature = signatureHex
+		tx.TransactionSignature = append(tx.TransactionSignature, rsvBytes)
+	}
+
+	return nil
+}
+
+func DecodeEcdsaSignature(signature xc.TxSignature) (btcec.ModNScalar, btcec.ModNScalar, error) {
```

Comment:
I think this is duplicated from elsewhere in the codebase?  i.e. the bitcoin tx package.

Please just important the relevant package and re-use it instead.  You could make a subpackage in bitcoin tx.